client/src/error.rs

```diff
@@ -0,0 +1,97 @@
+use std::str;
+
+use hyper;
+use serde_json;
+
+#[derive(Debug, Fail)]
+pub enum ClientError {
+    /// What was being attempted when the decode error occured.
+    // https://github.com/rust-lang-nursery/failure/issues/183 prevents
+    // easily deferring rendering until needed.
+    // For now, we render at instantiation as errors are relatively few
+    // we can amortise the costs well.
+    ///
+    /// description: Human summarised version of the failure.
+    /// The underlying failure is found in the first cause of the error chain.
+    /// The bytes that failed to decode.
+    /// TODO: Also capture the type that was being decoded into.
+    #[fail(display = "unable to parse {}: {}", description, _summary)]
+    DecodeFailed {
+        description: String,
+        _summary: String,
+        bytes: Vec<u8>,
+    },
+
+    #[fail(display = "Unexpected HTTP response status: {}", status)]
+    HttpStatusError { status: hyper::StatusCode },
+
+    #[fail(display = "Attribute {} required but not provided", attr)]
+    RequiredAttributeError { attr: &'static str },
+}
+
+impl ClientError {
+    pub fn new_decode_error(
+        description: &str,
+        e: &serde_json::Error,
+        bytes: Vec<u8>,
+    ) -> ClientError {
+        ClientError::DecodeFailed {
+            description: description.to_string(),
+            _summary: context(&e, &bytes),
+            bytes: bytes.clone(),
```

Comment:
`bytes` is already owned here, there's no need to clone it.

Reply:
This is actually in PR 5 - you probably want to review in order :)